Add options updating; Add session killing; Fix array parsing.

mshsmlv opened the pull request.

yalegko commented on session.go:
```diff
@@ -37,7 +39,13 @@ type Session struct {
-type Option func(cfg *SessionOptions)
+type ExistError struct {
+	name string
+}
```
Seems pretty useless with private `name` field and random name.

I suppose the intended usage is smth like
```go
if err != nil {
	var exists ExistError
	if errors.As(err, &exists) {
		etw.KillSession(exists.Name)
	}
}
```

yalegko commented on session.go:
```diff
@@ -37,7 +39,13 @@ type Session struct {
+}
+
+func (e ExistError) Error() string {
+	return fmt.Sprintf("session `%s` already exist", e.name)
```
`%q`?

yalegko commented on session.go:
```diff
@@ -111,6 +135,54 @@ func (s *Session) Close() error {
+	const (
+		maxLengthSessionName = 1024
+		maxLengthLogfileName = 1024
+	)
```
Why we need both if it was enough to allocate a space for name on connect?

And why we even need constants if we know length of the `name`?

mshsmlv replied:
We don't know if this session was opened with the log file or not. (session could be opened without our library)
That's why we allocate memory for LogFile name too.

I changed `maxLengthSessionName` to `len(name) * 2` -_0_0_-

yalegko replied:
Pls comment out the point about log file

__
> I changed maxLengthSessionName to len(name) * 2
```go
utf16Name, err := windows.UTF16FromString(s.config.Name)
...
l := len(utf16Name) * int(unsafe.Sizeof(utf16Name[0]))
```

yalegko commented on session.go:
```diff
@@ -143,8 +215,12 @@ func (s *Session) createETWSession() error {
+	}
+	if !errors.Is(err, windows.ERROR_SUCCESS) {
+		return fmt.Errorf("StartTrace failed; %w", err)
 	}
```
There is no errors wrapping here, so probably a simple switch would be more readable

yalegko commented on session.go:
```diff
@@ -90,6 +98,22 @@ func (s *Session) SubscribeAndServe() error {
+	}
+	if err := s.subscribeToProvider(); err != nil {
+		return err
+	}
```
Is it safe to call `.UpdateOptions` before the actual session start?

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,109 @@
+
+const (
+	MicrosoftWindowsDNSClient = "{1C95126E-7EEA-49A9-A3FE-A378B03DDB4D}"
+)
```
could be a private constant

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,109 @@
+	// Trying to create new session with the same name.
+	// Then stop the session properly.
+	// Ensure `listen` goroutine wont block forever.
+	////////////////////////////////////////////////////
```
Comments looks a bit outdated

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,109 @@
+	////////////////////////////////////////////////////
+	_, err = NewSession(guid, WithName(sessionName))
+	var exists *ExistError
+	if !errors.As(err, &exists) {
```
Should be checked `!= nil` also

mshsmlv replied:
Why? if `err` is nil `errors.As()` will return false also.

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,109 @@
+				t.Fatalf("Failed to close session properly; %s", err)
+			}
+		})
+	}
```
Stop in callback could be a bad idea if we can't (for some cases) recv an event. Then we'll never close the session and it will leak OS resources.

Same pattern as with `.Process` routine could be safer

```go
gotEvent := make(chan struct{}, 1)
cb := func(e *Event) {
	select {
	case gotEvent<-struct{}{}:
	default:
	}
}
```

yalegko commented on event.go:
```diff
@@ -267,24 +267,39 @@ func (p *propertyParser) getPropertyName(i int) string {
 
-// TODO: Return map[string]string
-func (p *propertyParser) parseComplexType(i int) ([]string, error) {
+func (p *propertyParser) parseComplexType(i int) (map[string]interface{}, error) {
```
`parseComplexType` -> `parseStruct`?

yalegko commented on event.go:
```diff
@@ -267,24 +267,39 @@ func (p *propertyParser) getPropertyName(i int) string {
-	return p.parseSimpleType(i)
+
+	return result, nil
 }
```
Didn't catch the idea. Why we return an array on any case? 

It seems pretty complex to use. Could we detect a real arrays using some API and return other fields as values?

mshsmlv replied:
Now it returns a value if array length is 1.

yalegko commented on go.sum:
```diff
@@ -1,2 +1,16 @@
+golang.org/x/sys v0.0.0-20180905080454-ebe1bf3edb33/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
+golang.org/x/sys v0.0.0-20190507160741-ecd444e8653b/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
 golang.org/x/sys v0.0.0-20200302150141-5c8b2ff67527 h1:uYVVQ9WP/Ds2ROhcaGPeIdVq0RIXVLwsHlnvJ+cT1So=
 golang.org/x/sys v0.0.0-20200302150141-5c8b2ff67527/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
```
`go mod tidy`?

mshsmlv replied:
Yes, I did.

yalegko commented on session.c:
```diff
@@ -83,6 +83,28 @@ int PropertyIsStruct(PTRACE_EVENT_INFO info, int i) {
+        TdhGetPropertySize(event, 0, NULL, 1, &dataDescriptor, &propertySize);
+        TdhGetProperty(event, 0, NULL, 1, &dataDescriptor, propertySize, (PBYTE)&count);
+        return count;
+    }
```
Could we re-use some code from `GetPropertyLength`? Looks almost similar

yalegko commented on session.c:
```diff
@@ -83,6 +83,28 @@ int PropertyIsStruct(PTRACE_EVENT_INFO info, int i) {
 
+// This is a copy-paste from MSDN.
+// https://docs.microsoft.com/ru-ru/windows/win32/etw/using-tdhformatproperty-to-consume-event-data
+int GetArraySize(PEVENT_RECORD event, PTRACE_EVENT_INFO info, int i) {
```
Move up to the `GetPropertyLength` please

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+			}
+		}
+	}
+}
```
Could we just generate a single event synchronously on demand? 

Seems it would be easier and more reliable to use in tests.

mshsmlv replied:
Event processing starts with a little delay. When event processing is ready to consume events, we have already sent the event.

That's why I generate a lot of events in a goroutine.

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+	case <-gotEvent:
+		// Pass.
+	case <-time.After(deadline):
+		p.Fail("Failed to receive event from provider in 10s")
```
`in 10s")` -> `in %s", deadline)`

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+	}
+}
+
+func (p *testProvider) TestSession() {
```
`TestSession_StartStop`?

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+
+	gotCriticalEvent := make(chan struct{}, 1)
+	gotInformationEvent := make(chan struct{}, 1)
+	cb := func(e *Event) {
```
Suppose some comments describing the testing scenario would be really helpful here

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+		case <-time.After(deadline):
+			p.Fail("Failed to get event with critical level")
+		}
+	}()
```
Why not just 2 synchronous blocks? Channels are bufferized anyway

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+		// Pass.
+		close(gotCriticalEvent)
+		close(gotInformationEvent)
+	case <-time.After(deadline):
```
Why we need to close channels here?

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+	// Ensure `listen` goroutine wont block forever.
+	if err := KillSession(sessionName); err != nil {
+		p.Fail("Failed to force session to stop %s", err)
+	}
```
`p.Require.NoError(err, "Failed to force stop session")`?

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+	}
+
+	// Killing the session by name.
+	// Ensure `listen` goroutine wont block forever.
```
>  	// Ensure `listen` goroutine wont block forever.
Outdated comment?

yalegko commented on session_test.go:
```diff
@@ -0,0 +1,313 @@
+			etw.Float64Field("float64", 46.7),
+			etw.Struct("subStructure",
+				etw.StringField("string", "string value"))),
+	)
```
How about array of structs? :)

mshsmlv replied:
Microsoft etw doesn't have function `StructArray`  :(

mshsmlv committed:
Add new API methods; Change API a little

NEW:
- Add method for updating session options on runtime
- Add method for force killing session by name
- Fix parsing for array-fields
- Cover API methods with test

Changes:
- Rename `SubscribeAndServe` -> Process
- Delete synchronize logic from API methods. Now it is the user responsible.

yalegko reviewed (approved):
LGTM
